pkg/relayer/session/claim.go

```diff
@@ -231,3 +256,83 @@ func (rs *relayerSessionsManager) goCreateClaimRoots(
 	failSubmitProofsSessionsCh <- failedClaims
 	claimsFlushedCh <- flushedClaims
 }
+
+// isProofRequired determines whether a proof is required for the given session's
+// claim based on the current proof module governance parameters.
+func (rs *relayerSessionsManager) isProofRequired(
+	ctx context.Context,
+	sessionTree relayer.SessionTree,
+) (isProofRequired bool, err error) {
+	logger := rs.logger.With(
+		"session_id", sessionTree.GetSessionHeader().GetSessionId(),
+		"claim_root", fmt.Sprintf("%x", sessionTree.GetClaimRoot()),
+		"supplier_operator_address", sessionTree.GetSupplierOperatorAddress().String(),
+	)
+
+	// Create the claim object and use its methods to determine if a proof is required.
+	claim := claimFromSessionTree(sessionTree)
+
+	// Get the number of compute units accumulated through the given session.
+	numClaimComputeUnits, err := claim.GetNumComputeUnits()
+	if err != nil {
+		return false, err
+	}
+
+	proofParams, err := rs.proofQueryClient.GetParams(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	logger = logger.With(
+		"num_claim_compute_units", numClaimComputeUnits,
+		"proof_requirement_threshold", proofParams.GetProofRequirementThreshold(),
+	)
+
+	// Require a proof if the claim's compute units meets or exceeds the threshold.
+	if numClaimComputeUnits >= proofParams.GetProofRequirementThreshold() {
+		logger.Info().Msg("compute units is above threshold, claim requires proof")
+
+		return true, nil
+	}
+
+	// Get the hash of the claim to seed the random number generator.
+	var claimHash []byte
+	claimHash, err = claim.GetHash()
+	if err != nil {
+		return false, err
+	}
+
+	// Sample a pseudo-random value between 0 and 1 to determine if a proof is required probabilistically.
+	var randFloat float32
+	randFloat, err = poktrand.SeededFloat32(claimHash[:])
+	if err != nil {
+		return false, err
+	}
```

Comment:
**⚠️  CRITICAL ⚠️** 

When I saw this, it raised a critical issue that we overlooked.

If the `rand(seed(claimHash))` is all we need to determine the probability, anyone can grind different claims until you don't need a proof.

In the same way that we use the block hash AFTER the claim closes (see `GetPathForProof`), we HAVE to do the same thing when computing this probability as well.

Reply:
Can you add a TODO_BETA in this PR but immediately tend to the correct business logic in another?

Reply:
_TL;DR; Seed should use more than claim hash, this has to be done both on/off-chain, will send a fix_

> re-creating the claim object

Actually no, we already have all we need to create a claim. There's nothing missing to do it exactly the same way on-chain does, see  [x/proof/keeper/msg_server_create_claim.go#L55](https://github.com/pokt-network/poktroll/blob/main/x/proof/keeper/msg_server_create_claim.go#L55)

>  we ONLY know if a proof is required AFTER

We also have all we need to know whether a poof is required or not, this flow reproduces how it's actually done on-chain, see: [main/x/tokenomics/keeper/settle_pending_claims.go#L282](https://github.com/pokt-network/poktroll/blob/main/x/tokenomics/keeper/settle_pending_claims.go#L282)

I agree that the actor can leverage the claim's root hash to get a seed that does not require a proof.

In order to fix that, that seed needs to include something that is not know as of the time of the claim submission.

If we add `ProofWindowOpenHeight` block hash to the random seed, we'll enforce that the probabilistic proof requirement is impossible to calculate before observing `ProofWindowOpenHeight`.

Which in turn will force the `RelayMiner` to push this logic to the proof logic.

If you agree with `ProofWindowOpenHeight` being the future value to use, I'll update both this logic and the on-chain part that enforces it.

Reply:
> TL;DR; Seed should use more than claim hash, this has to be done both on/off-chain, will send a fix

🤝  Agreed. Let's do it.

> There's nothing missing to do it...

 ❓  Can we create a shared helper?
 
 ---
 
 @red-0ne W.r.t to the rest, we're on the same page and aligned. WHat I was calling out is to following the exact same pattern as in `proof_validation.go`:
 
 ```go
 	// earliestSupplierProofCommitHeight - 1 is the block that will have its hash used as the
	// source of entropy for all the session trees in that batch, waiting for it to
	// be received before proceeding.
	proofPathSeedBlockHash := k.sessionKeeper.GetBlockHash(ctx, earliestSupplierProofCommitHeight-1)

	// TODO_BETA: Investigate "proof for the path provided does not match one expected by the on-chain protocol"
	// error that may occur due to block height differing from the off-chain part.
	k.logger.Info("E2E_DEBUG: height for block hash when verifying the proof", earliestSupplierProofCommitHeight, sessionHeader.GetSessionId())

	expectedProofPath := protocol.GetPathForProof(proofPathSeedBlockHash, sessionHeader.GetSessionId())
	if !bytes.Equal(proof.Path, expectedProofPath) {
		return types.ErrProofInvalidProof.Wrapf(
			"the path of the proof provided (%x) does not match one expected by the on-chain protocol (%x)",
			proof.Path,
			expectedProofPath,
		)
	}
```